nixos/modules/services/hardware/pid-fan-controller.nix

```
@@ -0,0 +1,190 @@
+{
+  lib,
+  config,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.pid-fan-controller;
+  heatSource = {
+    options = {
+      name = lib.mkOption {
+        type = lib.types.uniq lib.types.nonEmptyStr;
+        description = "Name of the heat source";
+      };
+      wildcardPath = lib.mkOption {
+        type = lib.types.nonEmptyStr;
+        description = ''
+          Path of the heat sources hwmon 'temp_input' file.
+          This path can contain multiple wildcards, but has to resolve to
+          exactly one result.
+        '';
+      };
+      pidParams = {
+        setPoint = lib.mkOption {
+          type = lib.types.ints.unsigned;
+          description = "Set point of the controller in °C";
+        };
+        P = lib.mkOption {
+          description = "K_p of PID controller";
+          type = lib.types.float;
+        };
+        I = lib.mkOption {
+          description = "K_i of PID controller";
+          type = lib.types.float;
+        };
+        D = lib.mkOption {
+          description = "K_d of PID controller";
+          type = lib.types.float;
+        };
+      };
+    };
+  };
+
+  fan = {
+    options = {
+      wildcardPath = lib.mkOption {
+        type = lib.types.str;
+        description = ''
+          Wildcard path of the hwmon 'pwm' file.
+          If the fans are not to be found in /sys/class/hwmon/hwmon* the corresponding
+          kernel module (like 'nct6775') needs to be added to `boot.kernelModules`.
+          See: [hwmon](https://www.kernel.org/doc/html/latest/hwmon/index.html)
+        '';
+      };
+      minPwm = lib.mkOption {
+        default = 0;
+        type = lib.types.ints.u8;
+        description = "Minimum PWM value";
+      };
+      maxPwm = lib.mkOption {
+        default = 255;
+        type = lib.types.ints.u8;
+        description = "Maximum PWM value";
+      };
+      cutoff = lib.mkOption {
+        default = false;
+        type = lib.types.bool;
+        description = "Whether to stop the fan when `minPwm` is reached";
+      };
+      heatPressureSrcs = lib.mkOption {
+        type = lib.types.nonEmptyListOf lib.types.str;
+        description = "Heat pressure sources affected by the fan";
```

```suggestion
        description = "List of heat pressure sources affected by the fan.";
```

An example on how to set this option would be nice.